lib/internal/modules/esm/loader.js

```diff
@@ -198,32 +204,46 @@ class ESMLoader {
    * user-defined loaders (as returned by ESMLoader.import()).
    */
   async addCustomLoaders(
-    customLoaders = [],
+    customLoaders = new SafeMap(),
   ) {
-    if (!ArrayIsArray(customLoaders)) customLoaders = [customLoaders];
-
-    for (let i = 0; i < customLoaders.length; i++) {
-      const exports = customLoaders[i];
+    // Maps are first-in-first-out, but hook chains are last-in-first-out,
+    // so create a new container for the incoming hooks (which have already
+    // been reversed).
+    const globalPreloaders = new SafeMap();
+    const resolvers = new SafeMap();
+    const loaders = new SafeMap();
+
+    for (const [url, exports] of customLoaders.entries()) {
       const {
         globalPreloader,
         resolver,
         loader,
       } = ESMLoader.pluckHooks(exports);
 
-      if (globalPreloader) ArrayPrototypePush(
-        this.#globalPreloaders,
+      if (globalPreloader) globalPreloaders.set(
+        url,
         FunctionPrototypeBind(globalPreloader, null), // [1]
       );
-      if (resolver) ArrayPrototypePush(
-        this.#resolvers,
+      if (resolver) resolvers.set(
+        url,
         FunctionPrototypeBind(resolver, null), // [1]
       );
-      if (loader) ArrayPrototypePush(
-        this.#loaders,
+      if (loader) loaders.set(
+        url,
         FunctionPrototypeBind(loader, null), // [1]
       );
     }
 
+    // Append the pre-existing hooks (the builtin/default ones)
+    for (const [k, v] of this.#globalPreloaders) globalPreloaders.set(k, v);
```

Comment:
Using array destructuring syntax over an array relies on the globally mutable `Array.prototype[@@iterator]` (I know, I know), you can use either
```suggestion
    for (const { 0: k, 1: v } of this.#globalPreloaders) globalPreloaders.set(k, v);
```
```suggestion
    for (const p of this.#globalPreloaders) globalPreloaders.set(p[0], p[1]);
```

Reply:
This is a `SafeMap`. Is that not safe from `Array.prototype[@@iterator]`? (I assumed it was 😓)

Reply:
It's safe to iterate over it, but you receive a genuine array object which itself is not safe to iterate over.

Reply:
Ah! Good to know, thank you!

Reply:
Is `SafeArrayIterator` safe for that? (I assumed `SafeMap`'s iterator returned that and that that was safe—that it was all kept in the "Safe" space).

Reply:
It's only "safe" one level deep, I remember trying to make it safe two level deep, but that was so not worth it performance-wise so I gave it up. You could use `SafeArrayIterator` here, but I wouldn't recommend it as I find it harder to read and would be less performant.
```suggestion
    for (const array of this.#globalPreloaders) {
      const [k, v] = new SafeArrayIterator(array);
      globalPreloaders.set(k, v);
    }
```